DigitalLearningSolutions.Data/Services/PasswordResetService.cs

```diff
@@ -1,120 +1,130 @@
-                            Note that this link can only be used once and it will expire in two hours.
-                            Please don’t reply to this email as it has been automatically generated.",
-                HtmlBody = $@"<body style= 'font - family: Calibri; font - size: small;'>
-                                    <p>Dear {firstName},</p>
-                                    <p>A request has been made to reset the password for your Digital Learning Solutions account.</p>
-                                    <p>To reset your password please follow this link: <a href=""{resetPasswordUrl.Uri}"">{resetPasswordUrl.Uri}</a></p>
-                                    <p>Note that this link can only be used once and it will expire in two hours.</p>
-                                    <p>Please don’t reply to this email as it has been automatically generated.</p>
-                                </body>"
-            };
-
-            return new Email(emailSubject, body, emailAddress);
-        }
-    }
-}
+﻿namespace DigitalLearningSolutions.Data.Services
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using DigitalLearningSolutions.Data.DataServices;
+    using DigitalLearningSolutions.Data.Enums;
+    using DigitalLearningSolutions.Data.Exceptions;
+    using DigitalLearningSolutions.Data.Models.Auth;
+    using DigitalLearningSolutions.Data.Models.Email;
+    using DigitalLearningSolutions.Data.Models.User;
+    using Microsoft.Extensions.Logging;
+    using MimeKit;
+
+    public interface IPasswordResetService
+    {
+        Task<bool> PasswordResetHashIsValidAsync(string emailAddress, string resetHash);
+        void GenerateAndSendPasswordResetLink(string emailAddress, string baseUrl);
+    }
+
+    public class PasswordResetService : IPasswordResetService
+    {
+        private readonly IPasswordResetDataService passwordResetDataService;
+        private readonly IEmailService emailService;
+        private readonly IClockService clockService;
+        private readonly ILogger<PasswordResetService> logger;
+
+        private readonly IUserService userService;
+
+        public PasswordResetService(
+            IUserService userService,
+            IPasswordResetDataService passwordResetDataService,
+            ILogger<PasswordResetService> logger,
+            IEmailService emailService,
+            IClockService clockService)
+        {
+            this.userService = userService;
+            this.passwordResetDataService = passwordResetDataService;
+            this.logger = logger;
+            this.emailService = emailService;
+            this.clockService = clockService;
+        }
+
+        public async Task<bool> PasswordResetHashIsValidAsync(string emailAddress, string resetHash)
+        {
+            var (adminUsersWithEmail, delegateUsersWithEmail) = userService.GetUsersByEmailAddress(emailAddress);

```

Comment:
Do you think it's worth doing a join with reset password table when fetching the users to simplify some of the logic below?

Reply:
Oops, using EF too much has made me lazy! Yep, think that would make it speedier.